### DISCUSSION TITLE
7.5-rc1 SetEnvironmentVariable breaking change, feature or bug?

### DISCUSSION THREAD

**Discussion by nightroman** [PowerShell/PowerShell#24605] [Q&A] [answered]
The following code produces different results in 7.4.6 and 7.5-rc1.
This breaks some things. Is this a feature to stay or a bug to fix?

```powershell
$name = 'TEST_VAR2'
$value = $null
[Environment]::SetEnvironmentVariable($name, $value)

$result = $env:TEST_VAR2

# PS 7.4.6 True
# PS 7.5.0 False
$null -eq $result
```

If this is a feature to stay, then it introduces some inconsistencies.
This "similar" code produces the same result in both PS versions:

```powershell
$env:TEST_VAR2 = $null
$result = $env:TEST_VAR2

# True in all versions
$null -eq $result
```

**Comment by rhubarb-geek-nz** [comment]
This is about the difference between dotnet 8 and dotnet 9. ~~Given you are calling a CLR method directly (apart from AMSI logging) PowerShell should not affect the behaviour~~.

See [Support for empty environment variables](https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/9.0/empty-env-variable)

**Reply by nightroman** [reply]
> apart from AMSI logging

I am not sure what it is. There are cases when calling `[Environment]::SetEnvironmentVariable` is ether convenient (the variable name is a variable itself) or the only way (when the third parameter is used ~ Process, User, Machine).

> PowerShell should not affect the behaviour

Yes, it should not but it does because we pass null, not an empty string.

**Comment by iSazonov** [comment, marked as answer]
Now there is a bug in 7.5.0 - env variable is not removed by `[Environment]::SetEnvironmentVariable($name, $null)`.

**Reply by nightroman** [reply]
Yes, looks like it's the old good feature of PowerShell converting nulls to empty strings on several occasions.

This particular occasion is especially unfortunate, it may have noticeable impact on adopting PowerShell 7.5...

**Reply by jborean93** [reply]
.NET 9 introduced the ability to set environment variables with an empty string when in the past an empty string would unset the env variable. While you did use `$null` here as the value one quirk of using `$null` for a string typed argument in PowerShell is that PowerShell automatically converts it to an empty string. To truly pass null you have to use the constant `[NullString]::Value` so

```powershell
[Environment]::SetEnvironmentVariable($name, [NullString]::Value)
```

You can also do `$env:EnvVar = $null` or if you are using a variable for the var name `Set-Item "env:$name" $null`

**Reply by nightroman** [reply]
Thank you. I have already adjusted my PowerShell toolset for the incoming potential issues.

It would be useful if the PowerShell team addresses this somehow. At least in release notes, as a "behavioral breaking change" (as dotnet documentation does), with recommended solutions, so that not too experienced users know what to do.

I am not sure that the feature of converting nulls to empty strings is properly documented at all. For example this "Learn PowerShell deep dive" does not mention `[NullString]::Value` or related use cases that users might wanted to know:
[Everything you wanted to know about $null - PowerShell | Microsoft Learn](https://learn.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-null?view=powershell-7.4)

**Reply by jborean93** [reply]
> It would be useful if the PowerShell team addresses this somehow. At least in release notes, as a "behavioral breaking change" (as dotnet documentation does), with recommended solutions, so that not too experienced users know what to do.

It's probably a good  idea put somewhere but I'm unsure what their common practices are here. Keep in mind that while some people may be aware of these .NET changes there's always a chance that other things have changed their behaviour that PowerShell has no idea off. I do agree that keeping track of these known behaviour changes is a good thing and nice to give to people as a reference.

> I am not sure that the feature of converting nulls to empty strings is properly documented at all

I'm not aware of any official docs about it, would be a good issue to raise on the documentation repo https://github.com/MicrosoftDocs/PowerShell-Docs. It's a very niche scenario to come across as typically most APIs treat `null` and `""` strings as the same but there are a few out there that do not (this now being one of them). Just as an FYI this behaviour has been in PowerShell since at least v3 so it's not new.

The reason why PowerShell does it is that it's trying to treat a string similar to other value types where they cannot actually be null. For example

```powershell
# Int32 is a ValueType and cannot be $null. It's set to the
# default value (0) when $null is set to it
[int]$a = $null
$a.GetType()  # Int32
$a  # 0

# While a string is a reference type PowerShell treats it like a
# ValueType and defines a "default" value ("") that applies when
# being set to $null
[string]$b = $null
$b.GetType()  # String

# A true reference type can be null
[hashtable]$c = $null
$c.GetType()  # InvalidOperation: You cannot call a method on a null-valued expression.
```

**Reply by nightroman** [reply]
> The reason why PowerShell does it is that it's trying to treat a string similar to other value types where they cannot actually be null.

It looks like swimming against the current to me, string is not a value type.

**Reply by rhubarb-geek-nz** [reply]
> It's a very niche scenario to come across as typically most APIs treat `null` and `""` strings as the same but there are a few out there that do not (this now being one of them).

When you say API are you referring to PowerShell API or Dotnet API? I would suggest that in dotnet itself "" and null are very different things.

**Reply by jborean93** [reply]
> It looks like swimming against the current to me, string is not a value type.

It isn't but PowerShell decided long ago to treat it somewhat like it. It's immutable like value types and you have to go out of your way to actually make it null. Whether that was the right decision or not is not for me to decide.

> When you say API are you referring to PowerShell API or Dotnet API? I would suggest that in dotnet itself "" and null are very different things.

They are certainly different but in reality a lot of .NET APIs do something like the below where they equate null and an empty string (sometimes whitespace as well) as the same thing.

```csharp
public void Method(string arg)
{
    if (string.IsNullOrEmpty(arg))
    {
        ...
    }
}
```

That's not to say all of them like that as `SetEnvironmentVariable` is now one of the less common situations where null and an empty string actually mean something different.

PowerShell itself is also typically the same as you typically see people do a truthy or call the `IsNullOrEmpty` method as a check

```powershell
if ($stringValue) {
}
```

**Comment by rhubarb-geek-nz** [comment]
Just to confirm the behaviour. PowerShell treats the type string special when doing direct method invocation.

```
#!/usr/bin/env pwsh

Add-Type @'
public class Foo
{
  public static bool IsNull(string s) { return s==null; }
}
public class Bar
{
  public static bool IsNull(object s) { return s==null; }
}
'@

'Foo'
[Foo]::IsNull([NullString]::Value)
[Foo]::IsNull($null)
[Foo]::IsNull("")
[Foo]::IsNull("bar")
[Foo].GetMethod('IsNull',[System.Reflection.BindingFlags]'Static, Public').Invoke($null,[object[]](,$null))
'Bar'
[Bar]::IsNull([NullString]::Value)
[Bar]::IsNull($null)
[Bar]::IsNull("")
[Bar]::IsNull("bar")
[Bar].GetMethod('IsNull',[System.Reflection.BindingFlags]'Static, Public').Invoke($null,[object[]](,$null))
```

object and string do not give the same results.

```
Foo
True
False
False
False
True
Bar
False
True
False
False
True
```

So my assumption that "PowerShell should not affect the behaviour" is false.

**Reply by jborean93** [reply]
> Just to confirm the behaviour. PowerShell treats the type string special when doing direct method invocation.

It's not just method invocation, it happens whenever you try to cast to a string typed variable like a parameter or just `[string]$var`

```powershell
Function Test-Function {
    param([string]$Parameter)

    if ($null -eq $Parameter) {
        "Is null"
    }
    else {
        $Parameter.GetType().FullName
    }
}

Test-Function $null
# System.String

[string]$var = $null
$null -eq $var
# False
```

Basically whenever PowerShell is told to cast `$null` to a `string`, it will do `""` instead.
